File: Python/HappyNumber.py
from math import floor

class Solution:
    """
    Write an algorithm to determine if a number n is happy.

    A happy number is a number defined by the following process:

        Starting with any positive integer, replace the number by the sum of the squares of its digits.
        Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
        Those numbers for which this process ends in 1 are happy.

    Return true if n is a happy number, and false if not.
    """
    def isHappy(self, n: int) -> bool:
        """Hashmap approach using int-string conversion
        Speed: O(I^D) where I = number of itermediary numbers, D = average number of digits in a number
        Space: O(I) 

        Args:
            n (int): _description_

        Returns:
            bool: _description_
        """

        hashMap = dict()

        #maxIterations = 20
        #iterationCounter = 0

        while n != 1:
            #iterationCounter += 1

            #if iterationCounter >= maxIterations:
            #    return False

            print(n)

            nStr = str(n)

            n = sum( int(digitStr)**2 for digitStr in nStr )

            #if loop found
            if n in hashMap:
                return False
            else:
                hashMap[n] = 0

        return True

class Solution2:
    def isHappy(self, n: int) -> bool:
        """Hashmap approach using entirely integers
        Speed: O(I^D) where I = number of itermediary numbers, D = average number of digits in a number
        Space: O(I) 

        Args:
            n (int): _description_

        Returns:
            bool: _description_
        """

        hashMap = dict()

        #maxIterations = 20
        #iterationCounter = 0

        while n != 1:
            #iterationCounter += 1

            #if iterationCounter >= maxIterations:
            #    return False

            #print(n)

            summation = 0

            while n >= 1:

                #get the ones places digit
                onesPlaceDigit = n % 10

                #print(onesPlaceDigit)

                #sum square of digit
                summation += onesPlaceDigit**2

                #shift digits right
                n = floor(n / 10)

            n = summation 

            #if loop found
            if n in hashMap:
                return False
            else:
                hashMap[n] = 0

        return True
    
class Solution3:

    def advanceNumber(self, n: int) -> int:
        summation = 0
       
        while n >= 1:

            #get the ones places digit
            onesPlaceDigit = n % 10

            #sum square of digit
            summation += onesPlaceDigit**2

            #shift digits right
            n = floor(n / 10)
        
        return summation 

    def isHappy(self, n: int) -> bool:
        """Tortoise and Hare/Floyd Cycle Detection alg (double pointer approach) with just ints
        Speed: O(I^D) where I = number of itermediary numbers, D = average number of digits in a number
        Space: O(1)

        Args:
            n (int): _description_

        Returns:
            bool: _description_
        """

        slow = n
        fast = n

        while True:
            
            slow = self.advanceNumber(slow)
            fast = self.advanceNumber(fast)
            fast = self.advanceNumber(fast)
            if slow == fast:
                break
        
        return slow == 1

File: C#/HappyNumber.cs
/// <summary>
/// Write an algorithm to determine if a number n is happy. A happy number is a number defined by the following process:
/// Starting with any positive integer, replace the number by the sum of the squares of its digits.
/// Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
/// Those numbers for which this process ends in 1 are happy.
/// Return true if n is a happy number, and false if not.
/// </summary>
public class HappyNumber
{
    /// <summary>
    /// 
    /// Speed: O(M) (linear)
    /// Space: O(M) where M = number of intermediate numbers till loop or 1 found
    ///     could be O(1) if tortoise + hair alg used to find loop instead w/ 2 pntrs cross
    /// </summary>
    /// <param name="n"></param>
    /// <returns></returns>
    public bool IsHappy(int n) {

        Dictionary<int, bool> history = new Dictionary<int, bool>();

        while(true)
        {
           int sqrOfDigits = 0;

           Console.WriteLine(n);

            while(n > 0)
            {
                int remainder = n % 10;

                Console.WriteLine(remainder.ToString());

                int remainderSqrd = remainder*remainder;

                sqrOfDigits += remainderSqrd;

                double nDivTen = n / 10; //floor funct must take double, not a float

                //Console.WriteLine($"{n} div'd by 10 {nDivTen.ToString()}");

                n = (int)Math.Floor( nDivTen );

                Console.WriteLine($"New n is {n.ToString()}");
            }

            if( sqrOfDigits == 1)
            {
                return true;
            }

            //if havent already seen number
            if( !history.TryGetValue(sqrOfDigits, out bool value)  )
            {
                //add number to history
                history.Add(sqrOfDigits, true);

                n = sqrOfDigits;
            }
            //not happy number bc loop repeating 
            else
            {
                return false;
            }   
        }
        
    }
}
